llvm/lib/Target/X86/X86ExpandPseudo.cpp

```diff
@@ -613,6 +613,87 @@ bool X86ExpandPseudo::expandMI(MachineBasicBlock &MBB,
   case X86::CALL64m_RVMARKER:
     expandCALL_RVMARKER(MBB, MBBI);
     return true;
+  case X86::ADD32mi_ND:
+  case X86::ADD64mi32_ND:
+  case X86::SUB32mi_ND:
+  case X86::SUB64mi32_ND:
+  case X86::AND32mi_ND:
+  case X86::AND64mi32_ND:
+  case X86::OR32mi_ND:
+  case X86::OR64mi32_ND:
+  case X86::XOR32mi_ND:
+  case X86::XOR64mi32_ND:
+  case X86::ADC32mi_ND:
+  case X86::ADC64mi32_ND:
+  case X86::SBB32mi_ND:
+  case X86::SBB64mi32_ND: {
+    // It's possible for an EVEX-encoded legacy instruction to reach the 15-byte
+    // instruction length limit: 4 bytes of EVEX prefix + 1 byte of opcode + 1
+    // byte of ModRM + 1 byte of SIB + 4 bytes of displacement + 4 bytes of
+    // immediate = 15 bytes in total, e.g.
+    //
+    //  addq    $184, -96, %rax
+    //
+    // In such a case, no additional segment override prefix can be used. To
+    // resolve the issue, we split the “long” instruction into 2 instructions:
+    //
+    //  subq    $184, %fs:257(%rbx, %rcx), %rax
+    //
+    //  ->
+    //
+    //  movq %fs:257(%rbx, %rcx)，%rax
+    //  subq $184, %rax
+    int MemOpNo = X86::getFirstAddrOperandIdx(MI);
+    Register Segment = MI.getOperand(MemOpNo + X86::AddrSegmentReg).getReg();
+    if (Segment == X86::NoRegister)
```

Comment:
Also need to check the 0x67 prefix:
```
$ echo 'subq $184, 257(%ebx, %ecx), %rax' | llvm-mc --show-encoding
        .text
llvm-mc: /export/users/pengfeiw/llvm-project/llvm/lib/Target/X86/MCTargetDesc/X86MCCodeEmitter.cpp:1937: virtual void (anonymous namespace)::X86MCCodeEmitter::encodeInstruction(const llvm::MCInst &, SmallVectorImpl<char> &, SmallVectorImpl<llvm::MCFixup> &, const llvm::MCSubtargetInfo &) const: Assertion `CB.size() - StartByte <= 15 && "The size of instruction must be no longer than 15."' failed.
```

Reply:
We explained this in https://discourse.llvm.org/t/rfc-support-long-instruction-fixup-for-x86/76539

APX is supported only in 64-bit mode, where compiler only use 64-bit memory addressing (namely base and index register are 64-bit, no ADSIZE)

Reply:
We have a mode called x32. It's belong to 64-bit but uses 32-bit memory. Are you sure APX doesn't support it? Is it a decision from GCC too?

Reply:
Good point! I forgot about x32. Let me add the test for ADSIZE.

Reply:
Done